Introduce additional personalization tags to the Email Editor

yuliyan opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Add personalization tags for Shopper Country, Order Date, Order Items, Order Subtotal, Order Tax, Order Total, Payment Method, Payment URL, Store Name, Store Address.

Closes [WOOPLUG-3655](https://linear.app/a8c/issue/WOOPLUG-3655/add-personalization-tags-for-order-and-customer).

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

| Screenshot |
| ------ |
| ![1UfAfzkpzt32E9Ex](https://github.com/user-attachments/assets/8198e2fe-33ed-4790-a1b5-3ba19ce33e30) |


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

1. Enable the block-based email editor from WooCommerce → Settings → Advanced → Features → **Block Email Editor (alpha)**.
2. Open an email in the editor WooCommerce → Settings → **Emails**.
3. Confirm the new personalization tags are present in the "Personalization Tags" pop up (can be opened from the contextual toolbar, refer to the screenshot).
4. Edit the "New order" email and add the new personalization tags.
5. Create a new order and confirm the personalization tags are replaced with the expected values in the received email.

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [x] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [x] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Rename "Shopper" to "Customer" in personalization tags and introduce additional personalization tags for Customer Country, Order Date, Order Items, Order Subtotal, Order Tax, Order Total, Payment Method, Payment URL, Store URL, Store Name, Store Address.</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

yuliyan committed:
Introduce additional personalization tags to the Email Editor

yuliyan committed:
Add changelog entry

triple0t commented on plugins/woocommerce/src/Internal/EmailEditor/PersonalizationTagManager.php:
```diff
@@ -162,6 +287,43 @@ function ( array $context ): string {
+				__( 'Store Address', 'woocommerce' ),
+				'woocommerce/store-address',
+				__( 'Store', 'woocommerce' ),
+				function (): string {
+					$address_parts = array(
+						WC()->countries->get_base_address(),
+						WC()->countries->get_base_address_2(),
+						WC()->countries->get_base_city(),
+						WC()->countries->get_base_state(),
+						WC()->countries->get_base_postcode(),
+						WC()->countries->countries[ WC()->countries->get_base_country() ] ?? WC()->countries->get_base_country() ?? '',
+					);
+
+					$address = implode( "\n", array_filter( $address_parts ) );
+					return trim( $address );
+				},
```
This [method](https://github.com/woocommerce/woocommerce/blob/2afc098512d2f6782339ca0b34c9320c70acdb57/plugins/woocommerce/includes/class-wc-emails.php#L710-L735) does the same thing, so it might be better to reuse it.

yuliyan replied:
While implementing the tag, I came across this method, but since it was marked as private, I initially decided not to use it.
After your suggestion, I took a second look I can agree it would be better to reuse it for consistency. I've updated the method visibility to public and refactored the tag to make use of it.

triple0t reviewed (commented):
I tested this and it works well. 

I can see the new tags in the editor
![Screenshot 2025-04-28 at 1 08 01 PM](https://github.com/user-attachments/assets/6f28c101-8568-49ee-a53c-76c6f21ed83b)


The tags are correctly rendered in the sent email and the preview
Email client
![Screenshot 2025-04-28 at 1 09 25 PM](https://github.com/user-attachments/assets/75255959-2abe-47e8-9239-d2e0be237cf7)

Preview
![Screenshot 2025-04-28 at 1 09 47 PM](https://github.com/user-attachments/assets/a0cfcd50-8e0c-47b0-966c-64b09aca681d)

I can also update the Order Date format
Editor
![Screenshot 2025-04-28 at 1 11 42 PM](https://github.com/user-attachments/assets/ba6aec65-6b42-4b38-96a9-318badf00334)
Email client
![Screenshot 2025-04-28 at 1 11 48 PM](https://github.com/user-attachments/assets/3c3e455d-012a-4d4d-bfbe-6277530d166a)


**Feedback from me.**
1. This https://github.com/woocommerce/woocommerce/blob/09759c21c6358179550ddef08462a489b9c6510f/plugins/woocommerce/src/Internal/EmailEditor/PersonalizationTagManager.php file is now a bit too long. Can we shorten it or refactor it?
2. Can we add a tag for Store URL? This should link to the Woo store URL e.g `/shop/`
3. On Preview and "Send test email", the tag information is totally empty. I think it would be nice to reuse the dummy order object here. ![oMrnLS.png](https://github.com/user-attachments/assets/8bc3e64b-a1c3-421c-9689-c3e78066698a)


Well done.

pavel-mailpoet commented:
> Also, I believe @pavel-mailpoet would prefer that we switch from Shopper to Customer. I don't have a strong opinion on this, but it's part of the issue spec.

WooCommerce is using customers in the UI, if we use Shopper the UX will not be consistent. @triple0t do you think we should keep the word Shopper? https://d.pr/i/Wid6Cr

triple0t commented:
@pavel-mailpoet `Shopper` is used on the email listing page.
![Screenshot 2025-04-28 at 1 31 02 PM](https://github.com/user-attachments/assets/c63aa0ce-da02-406c-8c5f-dc4a388da751)

I'm fine with switching to `Customer`.

yuliyan committed:
Use Customer term as a replacement for Shopper

yuliyan committed:
Use WC_Emails::get_store_address for woocommerce/store-address

yuliyan committed:
Add Store URL tag

yuliyan committed:
Update changelog entry

yuliyan commented:
Thanks for the feedback, @triple0t!

> This https://github.com/woocommerce/woocommerce/blob/09759c21c6358179550ddef08462a489b9c6510f/plugins/woocommerce/src/Internal/EmailEditor/PersonalizationTagManager.php file is now a bit too long. Can we shorten it or refactor it?

I agree it's getting quite large. Do you have any particular ideas on how you'd like to optimize it?
I was considering separating the individual categories into different files to reduce the size, but I'm not sure if that would bring meaningful improvements overall. Open to suggestions!

> Can we add a tag for Store URL? This should link to the Woo store URL e.g /shop/
Thanks, added the tag.

> On Preview and "Send test email", the tag information is totally empty. I think it would be nice to reuse the dummy order object here.

I'd rather address this separately so that the PR stays within the scope of the linked issue and doesn't get too big.

----

@pavel-mailpoet @triple0t I've renamed the "Shopper" personalizaiton tags to "Customer" and updated the block email templates using those.

triple0t reviewed (approved):
Well done, @yuliyan.
Nice job updating the changelog as well. 

> I agree it's getting quite large. Do you have any particular ideas on how you'd like to optimize it?
I was considering separating the individual categories into different files to reduce the size, but I'm not sure if that would bring meaningful improvements overall. Open to suggestions!

I was thinking of this as well, and I believe separating by category will at least improve the readability of the file.

> Thanks for catching that I missed this one. It's added now.

👍🏽

 > I'd rather address this separately so that the PR stays within the scope of the linked issue and doesn't get too big.

I'm totally fine with that. Please create the issue in the backlog.

> I've renamed the "Shopper" personalizaiton tags to "Customer" and updated the block email templates using those.

Awesome. Thank you.

I'm assigning the PR back to you. Please feel free to merge it or implement the last feedback.

yuliyan committed:
Refactor PersonalizationTagManager registering tags

yuliyan commented:
@triple0t I've created an [issue to add the dummy data](https://linear.app/a8c/issue/WOOPLUG-4088/add-the-dummy-data-to-the-personalizer-in-the-email-preview) and addressed the PersonalizationTagManager refactoring in https://github.com/woocommerce/woocommerce/pull/57479/commits/b272968a82d3942e07132d6f677b5b750e2e25b8. Could you take a look when you get the chance?

yuliyan committed:
Fix lint errors

triple0t commented on plugins/woocommerce/src/Internal/EmailEditor/PersonalizationTags/OrderTagsProvider.php:
```diff
@@ -0,0 +1,161 @@
+					if ( ! isset( $context['order'] ) ) {
+						return '';
+					}
+					return $context['order']->get_payment_method_title();
```
Because we are using a type hint, we declare strict types, and this is WooCommerce (extensions can do a lot of weird things when data passes through [filters](https://github.com/woocommerce/woocommerce/blob/00fbe210ea5fecadb1a723a746139b445885445f/plugins/woocommerce/includes/abstracts/abstract-wc-data.php#L882)). I suggest ensuring the return of this method is a string. 
e.g 
```php
$context['order']->get_payment_method_title() ?? '';
```

yuliyan replied:
Updated it here and in a couple of other places where it seems to be a similar case.

triple0t commented on plugins/woocommerce/src/Internal/EmailEditor/PersonalizationTags/OrderTagsProvider.php:
```diff
@@ -0,0 +1,161 @@
+					if ( ! isset( $context['order'] ) ) {
+						return '';
+					}
+					return $context['order']->get_checkout_payment_url();
```
Same here. Please ensure the return data is a string.

triple0t commented on plugins/woocommerce/src/Internal/EmailEditor/PersonalizationTags/OrderTagsProvider.php:
```diff
@@ -0,0 +1,161 @@
+			)
+		);
+
+		$registry->register(
+			new Personalization_Tag(
+				__( 'Admin Order Note', 'woocommerce' ),
+				'woocommerce/admin-order-note',
+				__( 'Store', 'woocommerce' ),
+				function ( array $context ): string {
+					if ( isset( $context['wc_email'], $context['wc_email']->customer_note ) ) {
+						return wptexturize( $context['wc_email']->customer_note );
+					}
+					return '';
+				},
+			)
+		);
```
Let's move this to the StoreTagsProvider for now. We plan to remove the tag when the custom block is created

yuliyan replied:
Oops, that was not intentional. Fixed it.

triple0t reviewed (commented):
@yuliyan, Thank you for refactoring the file. It's smaller, easier to read, and we can now easily write test cases for each category.
Well done.

yuliyan committed:
Move Admin Order Note to Store tags

yuliyan committed:
Improve tags return statements

yuliyan commented:
@triple0t, I've addressed the items from your latest feedback. Could you check if everything is all right?

triple0t reviewed (approved):
LGTM. Thank you
